### PR TITLE
Documentation: Add a page that explains how to use cmake

### LINKED ISSUE #3105: Document the CGAL CMake API
## Issue Details

With CGAL 4.12, the CMake API of CGAL has been modernized. In CGAL-4.13 we should now document how to use CGAL with CMake.

*(Actually, I thought there was already an issue entry for that but I cannot find it.)*

Here is a draft: https://github.com/CGAL/cgal/wiki/How-to-use-CGAL-with-CMake-or-your-own-build-system

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

Adds page in the developer manual that explains how to create a CMakeLists to use with CGAL.
## Release Management

* Affected package(s):Manual
* Issue(s) solved (if any): fix #3105
* Link to compiled documentation (obligatory for small feature) 
[Getting Started ](https://cgal.geometryfactory.com/~mgimeno/doc/CMakeTuto/Manual/general_intro.html)
[the new page](https://cgal.geometryfactory.com/~mgimeno/doc/CMakeTuto/Manual/devman_create_and_use_a_cmakelist.html)

### CODE REVIEW

**Comment by lrineau** [comment]
In the section *Minimal Example Using Qt5*, could we imagine a way so that the `CMakeLists.txt` and `test.cpp` are real files, tested by the testsuite?

I would rename `test.cpp` to `test_cgal_qt5`, to avoir using `test_` as the name of the target (because `test` is reserved by CMake). That looks strange.

**Comment by maxGimeno** [comment]
I could probably simply show the draw_triangulation_2 files

**Comment by afabri** [comment]
@lrineau didn't you plan to modernize our use of CMake?

**Comment by lrineau** [comment]
> @lrineau didn't you plan to modernize our use of CMake?

I have already do a modernization, long ago (PR #1436, merged late 2017 for CGAL-4.12). The following CMake API is the result of it
```cmake
add_executable(test_cgal test_cgal.cpp)
target_link_libraries(test_ CGAL::CGAL CGAL::CGAL_Qt5 CGAL::CGAL_Core)
```
It uses:
  - plain CMake commands (`find_package`, `add_executable`, and `target_link_libraries`),
  - imported targets (`CGAL::CGAL`) used with `target_link_libraries` to import all the properties needed for a CGAL project (include path, third-party libraries, compiler options).

What is still missing is:
  - a pruning of all trace of the old code in the whole CMake codebase of CGAL,
  - and a renewed handling of optional third-party dependencies (PR #4356 is probably an intermediate step).

But none of that will change the documentation Maxime already wrote.

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+\note The CGAL targets define the following compiler flags :
+  - `-frounding-math` with gcc
+  - `/fp:strict /fp:except-` with MSVC
+
+\section secexample Minimal Example Using Qt5
+This section describes a minimal example of a program that uses CGAL and Qt5 for some GUI features.
+
+\subsection subcmake CMakeLists.txt
+\code
+cmake_minimum_required(VERSION 3.1)
```
Comment (lrineau) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:37]:
```suggestion
cmake_minimum_required(VERSION 3.1...3.17)
```

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/*!
+\page devman_create_and_use_a_cmakelist How to use CGAL with CMake
```
Comment (lrineau) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:2]:
```suggestion
\page devman_create_and_use_a_cmakelist How to use %CGAL with CMake
```
... to remove the automatic linking of `CGAL`.

**Comment by afabri** [comment]
You might put `CGAL_target_use_Eigen(<target>)`  also in the Eigen section in the "Essential 3rd ..." page. And the same for all the other third party libaries.
eigen ->  Eigen
Can you put some empty lines in the CMakeLists.txt example, for example before each comment,

**Comment by lrineau** [comment]
The part about `CGAL_target_use_Eigen` (and similar macros) is the one that might change in the future, in favor of imported targets, like `CGAL::Eigen_support`.

That would change:
```cmake
add_executable(test_cgal test_cgal.cpp)
target_link_libraries(test_cgal PUBLIC CGAL::CGAL)
CGAL_target_use_Eigen(test_cgal)
```
into:
```cmake
add_executable(test_cgal test_cgal.cpp)
target_link_libraries(test_cgal PUBLIC CGAL::CGAL CGAL::Eigen_support)
```
Cc @sgiraudot

**Comment by maxGimeno** [comment]
> You might put `CGAL_target_use_Eigen(<target>)` also in the Eigen section in the "Essential 3rd ..." page. And the same for all the other third party libaries.
I don't understand what you want me to do.

**Comment by sgiraudot** [comment]
> You might put `CGAL_target_use_Eigen(<target>)` also in the Eigen section in the "Essential 3rd ..." page. And the same for all the other third party libaries.

It's already here: https://cgal.geometryfactory.com/CGAL/doc/master/Manual/thirdparty.html#title8

**Comment by lrineau** [comment]
> > You might put `CGAL_target_use_Eigen(<target>)` also in the Eigen section in the "Essential 3rd ..." page. And the same for all the other third party libaries.
> 
> I don't understand what you want me to do.

Let's avoid duplicating the information. It will be harder to keep it up-to-date.

**Comment by afabri** [comment]
It would be documented systematically in the 3rd party page, and just one example in this example, so I would not call this duplication. What worries me more is that we document something now that you plan to change.

**Comment by afabri** [comment]
Ooops, Simon you are right. It is already in place. Sorry about that,

**Comment by sgiraudot** [comment]
I'm having a look for switching to imported target for third party (`CGAL::Eigen_support` and so on), so maybe don't waste time on documenting the functions right now, they'll most likely be replaced soon.

**Comment by sgiraudot** [comment]
Do we agree that the CMakeLists.txt of our examples/tests should now look like this?
```cmake
  find_package(LASLIB)
  if (TARGET CGAL::LASLIB_support)
    add_executable( read_las_example "read_las_example.cpp" )
    target_link_libraries(read_las_example CGAL::CGAL CGAL::LASLIB_support)
  else()
    message(STATUS "NOTICE : the LAS reader test requires LASlib and will not be compiled.")
  endif()
```

**Comment by lrineau** [comment]
> Do we agree that the CMakeLists.txt of our examples/tests should now look like this?
> 
> ```cmake
>   find_package(LASLIB)
>   if (TARGET CGAL::LASLIB_support)
>     add_executable( read_las_example "read_las_example.cpp" )
>     target_link_libraries(read_las_example CGAL::CGAL CGAL::LASLIB_support)
>   else()
>     message(STATUS "NOTICE : the LAS reader test requires LASlib and will not be compiled.")
>   endif()
> ```

If the `FindLASLIB.cmake` is provided by CGAL, that can be implemented easily. But what about third-party libraries for which the `Find<Package>` module is provided by CMake (`FindBoost`, for example), or third-party libraries that ship `<Package>Config` module? For such a third-party library, where will be the CMake code that creates the target `CGAL::<Package>_support`?

**Comment by sgiraudot** [comment]
As far as I can tell, for all the libraries for which I created the `CGAL_target_use_[lib]` functions, we already provide `Find[lib].cmake`. I don't really know for the others…

**Comment by lrineau** [comment]
For the moment, there is Boost and OpenGR:
  - there is a `FindBoost` module shipped by CMake, and a `BoostConfig` module in recent version of Boost (>= 1.70),
  - OpenGR is shipped with a `OpenGRConfig` module,

Recent versions of TBB also have a `TBBConfig` module. And, actually, I think we, as the CGAL project, should try to avoid shipping `Find<Package>` modules, as few as possible. The proposal seems to imply that we will have to ship then forever.

**Comment by sgiraudot** [comment]
How about with use-files? It's not the prettiest solution but I guess it could work like that:
```cmake
  find_package(LASLIB)
  include(${CGAL_LASLIB_support})
  if (TARGET CGAL::LASLIB_support)
    add_executable( read_las_example "read_las_example.cpp" )
    target_link_libraries(read_las_example CGAL::CGAL CGAL::LASLIB_support)
  else()
    message(STATUS "NOTICE : the LAS reader test requires LASlib and will not be compiled.")
  endif()

**Comment by afabri** [comment]
> And, actually, I think we, as the CGAL project, should try to avoid shipping `Find<Package>` modules, as few as possible

But is that not in a full contradiction with what we discuss here?

**Comment by lrineau** [comment]
> How about with use-files? It's not the prettiest solution but I guess it could work like that:
> 
> ```cmake
>   find_package(LASLIB)
>   include(${CGAL_LASLIB_support})
>   if (TARGET CGAL::LASLIB_support)
>     add_executable( read_las_example "read_las_example.cpp" )
>     target_link_libraries(read_las_example CGAL::CGAL CGAL::LASLIB_support)
>   else()
>     message(STATUS "NOTICE : the LAS reader test requires LASlib and will not be compiled.")
>   endif()
> ```

We could even write directly the name of the module, without the use of a variable:
```cmake
include(CGAL_LASLIB_support)
```

**Comment by lrineau** [comment]
> > And, actually, I think we, as the CGAL project, should try to avoid shipping `Find<Package>` modules, as few as possible
> 
> But is that not in a full contradiction with what we discuss here?

Why? We are trying to define CGAL specific imported targets, on top of whatever is already defined by the calls to `find_package(<package>)`. The issue is all the macros we use, to deal with optional dependencies.
```shellsession
> ack -- -D CGAL_target*
CGAL_target_use_Eigen.cmake
11:  target_compile_options( ${target} PUBLIC -DCGAL_EIGEN3_ENABLED)

CGAL_target_use_GLPK.cmake
8:  target_compile_options(${target} PUBLIC -DCGAL_USE_GLPK)

CGAL_target_use_LASLIB.cmake
9:  target_compile_options( ${target} PUBLIC -DCGAL_LINKED_WITH_LASLIB)

CGAL_target_use_OpenCV.cmake
8:  target_compile_options( ${target} PUBLIC -DCGAL_LINKED_WITH_OPENCV)

CGAL_target_use_OpenGR.cmake
8:  target_compile_options( ${target} PUBLIC -DCGAL_LINKED_WITH_OPENGR)

CGAL_target_use_pointmatcher.cmake
8:  target_compile_options( ${target} PUBLIC -DCGAL_LINKED_WITH_POINTMATCHER)

CGAL_target_use_SCIP.cmake
8:  target_compile_options(${target} PUBLIC -DCGAL_USE_SCIP)

CGAL_target_use_TBB.cmake
13:  target_compile_options( ${target} PUBLIC -DNOMINMAX -DCGAL_LINKED_WITH_TBB )

CGAL_target_use_TensorFlow.cmake
8:  target_compile_options( ${target} PUBLIC -DCGAL_LINKED_WITH_TENSORFLOW)

Ack finished at Wed Apr  8 18:06:00
```

**Comment by sgiraudot** [comment]
Just so we can move on, do we agree on this?
- targets are called `CGAL::[Lib]_support`*
- users must call `include(CGAL_[lib]_support)` after `find_package([lib])`

_* ` (`CGAL::[Lib]` was also suggested but it seems more vague to me, `CGAL::Eigen` sounds like we are providing Eigen or features for Eigen, whereas `CGAL::Eigen_support` seems to be clearer: we want to activate Eigen support in CGAL)_

**Comment by afabri** [comment]
fine with me

**Comment by sgiraudot** [comment]
I made a proposal here: https://github.com/CGAL/cgal/pull/4636

It's not finished: please have a look at it and tell me if this looks good. Then I'll make the changes everywhere (there are quite many occurrences, so I don't want to process the whole thing until I'm 100% sure that's the way we want it done :) ).

**Comment by maxGimeno** [comment]
success in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-121.shtml

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+\endcode
+
+Other CGAL libraries are linked similarly. For example, with CGAL_Core:
+
+\code
+find_package(CGAL REQUIRED COMPONENTS Core)
+target_link_libraries(my_executable CGAL::CGAL CGAL::CGAL_Core)
+\endcode
+
+There are also some cmake macros to link with CGAL dependenciesthat can be found in
```
Comment (sloriot) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:21]:
```suggestion
There are also some cmake macros to link with CGAL dependencies that can be found in
```

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+the section \subpage thirdparty.
+
+\note The CGAL targets define the following compiler flags :
+  - `-frounding-math` with gcc
+  - `/fp:strict /fp:except-` with MSVC
+
+\section secexample Minimal Example Using Qt5
+This section describes a minimal example of a program that uses %CGAL and Qt5 for some GUI features.
```
Comment (sloriot) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:29]:
```suggestion
This section describes a minimal example of a program that uses \cgal and Qt5 for some GUI features.
```

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 5]
```diff
@@ -0,0 +1,44 @@
+/*!
+\page devman_create_and_use_a_cmakelist How to use %CGAL with CMake
+
+This page will explain how to manually create a CMakeLists.txt file to link a custom program with CGAL.
+A base can be created using the script `cgal_create_CMakeLists`. Its usage is detailed in Section \ref devman_create_cgal_CMakeLists.
+\section seclink Linking with CGAL
```
Comment (sloriot) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:6]:
```suggestion
\section seclink Linking with \cgal
```

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 6]
```diff
@@ -0,0 +1,44 @@
+/*!
+\page devman_create_and_use_a_cmakelist How to use %CGAL with CMake
+
+This page will explain how to manually create a CMakeLists.txt file to link a custom program with CGAL.
```
Comment (sloriot) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:4]:
```suggestion
This page will explain how to manually create a CMakeLists.txt file to link a custom program with \cgal.
```

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/*!
+\page devman_create_and_use_a_cmakelist How to use %CGAL with CMake
```
Comment (sloriot) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:2]:
```suggestion
\page devman_create_and_use_a_cmakelist How to use \cgal with CMake
```

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+This page will explain how to manually create a CMakeLists.txt file to link a custom program with CGAL.
+A base can be created using the script `cgal_create_CMakeLists`. Its usage is detailed in Section \ref devman_create_cgal_CMakeLists.
+\section seclink Linking with CGAL
+To link with the CGAL library, use the following :
```
Comment (sloriot) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:7]:
```suggestion
To link with the \cgal library, use the following :
```

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt (hunk 4 above)
Comment (sloriot) [inline Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:29]:
```suggestion
This section describes a minimal example of a program that uses \cgal and Qt5 for some GUI features.
```

File: Documentation/doc/Documentation/Getting_started.txt [hunk 9]
```diff
@@ -9,6 +9,8 @@ The following pages describe how to use \cgal on different environments
 
 - \subpage thirdparty gives information (supported versions, download links) of the required and optional third party libraries.
 
+- \subpage devman_create_and_use_a_cmakelist explains how to use CMake to link a program with CGAL.
```
Comment (sloriot) [inline Documentation/doc/Documentation/Getting_started.txt:12]:
```suggestion
- \subpage devman_create_and_use_a_cmakelist explains how to use CMake to link a program with \cgal.
```

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-126.shtml

**Comment by sloriot** [comment]
I locally that the output was valid.

File: Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt (hunk 2 above)
Comment (lrineau) [reply Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:2]: @sloriot The initial issue (a link on "CGAL") is still there, now that you can reverted to using `CGAL`. I think it should be `%CGAL`.

See here: https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-5.1-Ic-127/output2/Manual/devman_create_and_use_a_cmakelist.html

Comment (sloriot) [reply Documentation/doc/Documentation/Developer_manual/create_and_use_a_cmakelist.txt:2]: I saw yesterday that it was breaking the navtree but actually it was another line. Fixed in 2dc7f66c8a5.

**Pull request merged**